File: app/src/main/java/com/winlator/core/ProcessHelper.java
package com.winlator.core;

import android.os.Process;
import android.util.Log;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Map;
import java.util.concurrent.Executors;

public abstract class ProcessHelper {
    public static final boolean PRINT_DEBUG = false; // FIXME change to false
    private static final ArrayList<Callback<String>> debugCallbacks = new ArrayList<>();
    private static final byte SIGCONT = 18;
    private static final byte SIGSTOP = 19;

    public static void suspendProcess(int pid) {
        Process.sendSignal(pid, SIGSTOP);
    }

    public static void resumeProcess(int pid) {
        Process.sendSignal(pid, SIGCONT);
    }

    public static int exec(String command) {
        return exec(command, null);
    }

    public static int exec(String command, String[] envp) {
        return exec(command, envp, null);
    }

    public static int exec(String command, String[] envp, File workingDir) {
        return exec(command, envp, workingDir, null);
    }

    public static int exec(String command, String[] envp, File workingDir, Callback<Integer> terminationCallback) {
        int pid = -1;
        try {
            Log.d("ProcessHelper", "Executing: " + Arrays.toString(splitCommand(command)) + ", " + Arrays.toString(envp) + ", " + workingDir);
            java.lang.Process process = Runtime.getRuntime().exec(splitCommand(command), envp, workingDir);
            // ProcessBuilder builder = new ProcessBuilder()
            //         .command(splitCommand(command))
            //         .directory(workingDir)
            //         .inheritIO();
            // // Add environment variables
            // if (envp != null) {
            //     Map<String, String> environment = builder.environment();
            //     for (String entry : envp) {
            //         String[] parts = entry.split("=", 2);
            //         if (parts.length == 2) {
            //             environment.put(parts[0], parts[1]);
            //         }
            //     }
            // }
            // java.lang.Process process = builder.start();

            Field pidField = process.getClass().getDeclaredField("pid");
            pidField.setAccessible(true);
            pid = pidField.getInt(process);
            pidField.setAccessible(false);

            if (!debugCallbacks.isEmpty()) {
                createDebugThread(process.getInputStream());
                createDebugThread(process.getErrorStream());
            }

            if (terminationCallback != null) createWaitForThread(process, terminationCallback);
        }
        catch (Exception e) {
            Log.e("ProcessHelper", "Failed to execute command: " + e);
        }
        return pid;
    }

    private static void createDebugThread(final InputStream inputStream) {
        Executors.newSingleThreadExecutor().execute(() -> {
            try (BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    if (PRINT_DEBUG) System.out.println(line);
                    synchronized (debugCallbacks) {
                        if (!debugCallbacks.isEmpty()) {
                            for (Callback<String> callback : debugCallbacks) callback.call(line);
                        }
                    }
                }
            }
            catch (IOException e) {
                Log.e("ProcessHelper", "Error on debug thread: " + e);
            }
        });
    }

    private static void createWaitForThread(java.lang.Process process, final Callback<Integer> terminationCallback) {
        Executors.newSingleThreadExecutor().execute(() -> {
            try {
                int status = process.waitFor();
                terminationCallback.call(status);
            }
            catch (InterruptedException e) {
                Log.e("ProcessHelper", "Error while waiting for thread: " + e);
            }
        });
    }

    public static void removeAllDebugCallbacks() {
        synchronized (debugCallbacks) {
            debugCallbacks.clear();
        }
    }

    public static void addDebugCallback(Callback<String> callback) {
        synchronized (debugCallbacks) {
            if (!debugCallbacks.contains(callback)) debugCallbacks.add(callback);
        }
    }

    public static void removeDebugCallback(Callback<String> callback) {
        synchronized (debugCallbacks) {
            debugCallbacks.remove(callback);
        }
    }

    public static String[] splitCommand(String command) {
        ArrayList<String> result = new ArrayList<>();
        boolean startedQuotes = false;
        String value = "";
        char currChar, nextChar;
        for (int i = 0, count = command.length(); i < count; i++) {
            currChar = command.charAt(i);
            char quoteChar = '"';

            if (startedQuotes) {
                if (currChar == quoteChar) {
                    startedQuotes = false;
                    if (!value.isEmpty()) {
                        value += quoteChar;
                        result.add(value);
                        value = "";
                    }
                }
                else value += currChar;
            }
            else if (currChar == '"' || currChar == '\'') {
                if (currChar == '\'') quoteChar = '\'';
                startedQuotes = true;
                value += quoteChar;
            }
            else {
                nextChar = i < count-1 ? command.charAt(i+1) : '\0';
                if (currChar == ' ' || (currChar == '\\' && nextChar == ' ')) {
                    if (currChar == '\\') {
                        value += ' ';
                        i++;
                    }
                    else if (!value.isEmpty()) {
                        result.add(value);
                        value = "";
                    }
                }
                else {
                    value += currChar;
                    if (i == count-1) {
                        result.add(value);
                        value = "";
                    }
                }
            }
        }

        return result.toArray(new String[0]);
    }

    public static String getAffinityMaskAsHexString(String cpuList) {
        String[] values = cpuList.split(",");
        int affinityMask = 0;
        for (String value : values) {
            byte index = Byte.parseByte(value);
            affinityMask |= (int)Math.pow(2, index);
        }
        return Integer.toHexString(affinityMask);
    }

    public static int getAffinityMask(String cpuList) {
        if (cpuList == null || cpuList.isEmpty()) return 0;
        String[] values = cpuList.split(",");
        int affinityMask = 0;
        for (String value : values) {
            byte index = Byte.parseByte(value);
            affinityMask |= (int)Math.pow(2, index);
        }
        return affinityMask;
    }

    public static int getAffinityMask(boolean[] cpuList) {
        int affinityMask = 0;
        for (int i = 0; i < cpuList.length; i++) {
            if (cpuList[i]) affinityMask |= (int)Math.pow(2, i);
        }
        return affinityMask;
    }

    public static int getAffinityMask(int from, int to) {
        int affinityMask = 0;
        for (int i = from; i < to; i++) affinityMask |= (int)Math.pow(2, i);
        return affinityMask;
    }
}

File: winlator/src/main/java/com/winlator/core/ProcessHelper.kt
package com.winlator.core

import java.io.BufferedReader
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.util.concurrent.Executors
import kotlin.math.pow
import timber.log.Timber

object ProcessHelper {
    const val PRINT_DEBUG: Boolean = false // FIXME change to false

    private val debugCallbacks = ArrayList<Callback<String>>()
    private const val SIGCONT: Byte = 18
    private const val SIGSTOP: Byte = 19

    fun suspendProcess(pid: Int) {
        android.os.Process.sendSignal(pid, SIGSTOP.toInt())
    }

    fun resumeProcess(pid: Int) {
        android.os.Process.sendSignal(pid, SIGCONT.toInt())
    }

    @JvmOverloads
    fun exec(
        command: String,
        envp: Array<String?>? = null,
        workingDir: File? = null,
        terminationCallback: Callback<Int>? = null,
    ): Int {
        var pid = -1

        try {
            val process = Runtime.getRuntime().exec(splitCommand(command), envp, workingDir)
            val pidField = process.javaClass.getDeclaredField("pid")

            pidField.isAccessible = true
            pid = pidField.getInt(process)
            pidField.isAccessible = false

            if (debugCallbacks.isNotEmpty()) {
                createDebugThread(process.inputStream)
                createDebugThread(process.errorStream)
            }

            if (terminationCallback != null) {
                createWaitForThread(process, terminationCallback)
            }
        } catch (e: Exception) {
            Timber.w(e)
        }

        return pid
    }

    private fun createDebugThread(inputStream: InputStream) {
        Executors.newSingleThreadExecutor().execute {
            try {
                BufferedReader(InputStreamReader(inputStream)).use { reader ->
                    var line: String
                    while ((reader.readLine().also { line = it }) != null) {
                        if (PRINT_DEBUG) {
                            println(line)
                        }

                        synchronized(debugCallbacks) {
                            if (debugCallbacks.isNotEmpty()) {
                                for (callback in debugCallbacks) {
                                    callback.call(line)
                                }
                            }
                        }
                    }
                }
            } catch (e: IOException) {
                Timber.w(e)
            }
        }
    }

    private fun createWaitForThread(process: Process, terminationCallback: Callback<Int>) {
        Executors.newSingleThreadExecutor().execute {
            try {
                val status = process.waitFor()
                terminationCallback.call(status)
            } catch (_: InterruptedException) {
            }
        }
    }

    fun removeAllDebugCallbacks() {
        synchronized(debugCallbacks) {
            debugCallbacks.clear()
        }
    }

    fun addDebugCallback(callback: Callback<String>) {
        synchronized(debugCallbacks) {
            if (!debugCallbacks.contains(callback)) debugCallbacks.add(callback)
        }
    }

    fun removeDebugCallback(callback: Callback<String>) {
        synchronized(debugCallbacks) {
            debugCallbacks.remove(callback)
        }
    }

    fun splitCommand(command: String): Array<String> {
        val result = ArrayList<String>()
        var startedQuotes = false
        var value = ""
        var currChar: Char
        var nextChar: Char
        var i = 0
        val count = command.length

        while (i < count) {
            currChar = command[i]

            if (startedQuotes) {
                if (currChar == '"') {
                    startedQuotes = false

                    if (value.isNotEmpty()) {
                        value += '"'
                        result.add(value)
                        value = ""
                    }
                } else {
                    value += currChar
                }
            } else if (currChar == '"') {
                startedQuotes = true
                value += '"'
            } else {
                nextChar = if (i < count - 1) command[i + 1] else '\u0000'

                if (currChar == ' ' || (currChar == '\\' && nextChar == ' ')) {
                    if (currChar == '\\') {
                        value += ' '
                        i++
                    } else if (!value.isEmpty()) {
                        result.add(value)
                        value = ""
                    }
                } else {
                    value += currChar

                    if (i == count - 1) {
                        result.add(value)
                        value = ""
                    }
                }
            }

            i++
        }

        return result.toTypedArray<String>()
    }

    fun getAffinityMaskAsHexString(cpuList: String): String {
        val values = cpuList.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        var affinityMask = 0

        for (value in values) {
            val index = value.toByte()
            affinityMask = affinityMask or 2.0.pow(index.toDouble()).toInt()
        }

        return Integer.toHexString(affinityMask)
    }

    fun getAffinityMask(cpuList: String?): Int {
        if (cpuList.isNullOrEmpty()) {
            return 0
        }

        val values = cpuList.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        var affinityMask = 0

        for (value in values) {
            val index = value.toByte()
            affinityMask = affinityMask or 2.0.pow(index.toDouble()).toInt()
        }

        return affinityMask
    }

    fun getAffinityMask(cpuList: BooleanArray): Int {
        var affinityMask = 0

        for (i in cpuList.indices) {
            if (cpuList[i]) affinityMask = affinityMask or 2.0.pow(i.toDouble()).toInt()
        }

        return affinityMask
    }

    fun getAffinityMask(from: Int, to: Int): Int {
        var affinityMask = 0

        for (i in from..<to) {
            affinityMask = affinityMask or 2.0.pow(i.toDouble()).toInt()
        }

        return affinityMask
    }
}
